state-chain/chains/src/sol/instruction_builder.rs

```diff
@@ -27,15 +26,15 @@ use crate::{
 };
 
 use super::{
-	api::{SolanaEnvironment, TokenEnvironment},
+	api::{get_token_decimals, SolanaEnvironment, TokenEnvironment},
 	consts::TOKEN_PROGRAM_ID,
 };
 use sp_std::str::FromStr;
 
 /// Internal enum type that contains SolAsset with derived ATA
-pub enum AssetWithDerivedAddress {
-	Sol,
-	Usdc(DerivedAta),
+pub struct AssetWithDerivedAddress {
+	pub asset: SolAsset,
+	pub ata: Option<DerivedAta>,
```

Comment:
I feel like the previous enum might have been a better choice than a struct. 

The struct doesn't prevent someone from building this for example:

```rust
AssetWithDerivedAddress {
	asset: Sol,
	ata: Some(ata)
}
```

which would be invalid.

Reply:
In fact I followed this code through to where it's used an I think `decompose_fetch_params` and `AssetWithDerivedAddress` are not necessary. 

We could just call `derive_associated_token_account` in `fetch_from` and then remove all this code.

Reply:
I changed it to make the ```ata``` available in the generic token_asset arm.

I also thought that ```decompose_fetch_params``` is not needed but didnt want to change it as part of this PR but I guess I'll just do it. makes it much simpler